homeassistant/components/screenlogic/sensor.py

```diff
@@ -85,190 +146,92 @@
     STATE_TYPE.TOTAL_INCREASING: SensorStateClass.TOTAL_INCREASING,
 }
 
-SL_UNIT_TO_HA_UNIT = {
-    UNIT.CELSIUS: UnitOfTemperature.CELSIUS,
-    UNIT.FAHRENHEIT: UnitOfTemperature.FAHRENHEIT,
-    UNIT.MILLIVOLT: UnitOfElectricPotential.MILLIVOLT,
-    UNIT.WATT: UnitOfPower.WATT,
-    UNIT.HOUR: UnitOfTime.HOURS,
-    UNIT.SECOND: UnitOfTime.SECONDS,
-    UNIT.REVOLUTIONS_PER_MINUTE: REVOLUTIONS_PER_MINUTE,
-    UNIT.PARTS_PER_MILLION: CONCENTRATION_PARTS_PER_MILLION,
-    UNIT.PERCENT: PERCENTAGE,
-}
-
 
 async def async_setup_entry(
     hass: HomeAssistant,
     config_entry: ConfigEntry,
     async_add_entities: AddEntitiesCallback,
 ) -> None:
     """Set up entry."""
-    entities: list[ScreenLogicSensorEntity] = []
-    coordinator: ScreenlogicDataUpdateCoordinator = hass.data[DOMAIN][
+    entities: list[ScreenLogicSensor] = []
+    coordinator: ScreenlogicDataUpdateCoordinator = hass.data[SL_DOMAIN][
         config_entry.entry_id
     ]
-    equipment_flags = coordinator.gateway_data[SL_DATA.KEY_CONFIG]["equipment_flags"]
 
-    # Generic push sensors
-    for sensor_name in coordinator.gateway_data[SL_DATA.KEY_SENSORS]:
-        if sensor_name in SUPPORTED_BASIC_SENSORS:
-            entities.append(
-                ScreenLogicStatusSensor(coordinator, sensor_name, CODE.STATUS_CHANGED)
-            )
+    for base_kwargs, base_data in process_supported_values(
+        coordinator, DOMAIN, SUPPORTED_DATA
+    ):
+        base_kwargs["device_class"] = SL_DEVICE_TYPE_TO_HA_DEVICE_CLASS.get(
+            base_data.value_data.get(ATTR.DEVICE_TYPE)
+        )
+        base_kwargs["native_unit_of_measurement"] = get_ha_unit(base_data.value_data)
+        base_kwargs["options"] = base_data.value_data.get(ATTR.ENUM_OPTIONS)
+        base_kwargs["state_class"] = SL_STATE_TYPE_TO_HA_STATE_CLASS.get(
+            base_data.value_data.get(ATTR.STATE_TYPE)
+        )
+        base_kwargs["value_mod"] = base_data.value_parameters.get(
+            EntityParameter.VALUE_MODIFICATION
+        )
```

Comment:
There is so much going on here that its going to take me a few days to wrap my head around all of this as there is a lot of code to trace and review here.  I was hoping the final form was going to be a bit less complex.

Ideally we come up with a way to generate ScreenLogicSensorEntityDescription objects ahead of time and filter them down to what is supported at run time instead of constructing them all at run time.

Reply:
What point is considered "ahead of time" and not "run time"?

Reply:
Ah ok. I wasn't thinking about it in that way. My last changes to `data` are not really on the right track then, but I think I understand now. 

Back to the drawing board!

Reply:
I've tried to make the configurable entities more follow the pattern of `switch`, with entity creation logic in each platform setup. 

This keeps most HA stuff out of `data`, except for `cleqnup_excluded_entity`, which is isolated so I could relocate it somewhere if desired.